File: ex13slice.m
% ex13: Slicing, MATLAB

foo = [11 22 33 44];

% Slicing makes copies
bar = foo(2:3);  % bar=[22 33]
bar(1) = 0;  % foo unchanged

baz = foo(2:end);  %#ok, Slice to end
baz = foo(end);  %#ok, Last value: "end means last one"
baz = foo(end-1); % Backwards indexing

disp([foo bar baz]);

foo = [1 2 3; 4 5 6];

% Colon alone, "all"
bar = foo(:, 2);  %#ok,  2 5
bar = foo(1, :);  %#ok,  1 2 3
bar = foo(:);  %  1x6: 1 2 3 4 5 6

% Partial indexes
baz = foo(1);  %  1

disp([foo bar baz]);

File: ex13slice.py
# ex13: Slicing, MATLAB

import numpy as np

foo = np.array([11, 22, 33, 44])

# Slicing creates "view" into orig
bar = foo[1:2]  # bar=[22, 33]
bar[0] = 0  # foo=[0, 22. 33, 44], bar=[0, 33, 44]  !!!

baz = foo[1:]  # Slice to end
baz = foo[-1]  # Last value: "negative means wrap around"
baz = foo[-2]  # Backwards indexing

print(foo, bar, baz)

foo = np.array([[1, 2, 3], [4, 5, 6]])

# Colon alone, "all"
bar = foo[:, 1]  # [2, 5]
bar = foo[0, :]  # [1, 2, 3]
bar = foo[:]  # 2x3: [[1, 2, 3], [4, 5, 6]]  !!!

# Partial indexes
baz = foo[0]  # [1, 2, 3] !!!

print(foo, bar, baz)
